lib/macs/cbc_block_cipher_mac.dart

```diff
@@ -0,0 +1,210 @@
+     * Cryptography).
+     *
+     * @param cipher the cipher to be used as the basis of the MAC generation.
+     * @param macSizeInBits the size of the MAC in bits, must be a multiple of 8.
+     * @param padding the padding to be used to complete the last block.
+     */
+  CBCBlockCipherMac(BlockCipher cipher, int macSizeInBits, Padding padding)
+      : _cipher = new CBCBlockCipher(cipher),
+        _macSize = macSizeInBits ~/ 8,
+        _padding = padding {
+    if ((macSizeInBits % 8) != 0) {
+      throw new ArgumentError("MAC size must be multiple of 8");
+    }
+
+    _mac = new Uint8List(cipher.blockSize);
+
+    _buf = new Uint8List(cipher.blockSize);
+    _bufOff = 0;
+  }
+
+  @override
+  String get algorithmName {
+    String paddingName = _padding != null ? "/${_padding.algorithmName}" : "";
+    return "${_cipher.algorithmName}_CMAC${paddingName}";
+  }
+
+  @override
+  void init(covariant KeyParameter keyParams) {
+    final zeroIV = new Uint8List(keyParams.key.length);
+    this._params = new ParametersWithIV(keyParams, zeroIV);
+
+    reset();
+
+    _cipher.init(true, params);
+  }
+
+  @override
+  get macSize => _macSize;
+
+  @override
+  void updateByte(int inp) {
+    if (_bufOff == _buf.length) {
+      _cipher.processBlock(_buf, 0, _mac, 0);
+      _bufOff = 0;
+    }
+
+    _buf[_bufOff++] = inp;
+  }
+
+  @override
+  void update(Uint8List inp, int inOff, int len) {
+    if (len < 0) {
+      throw new ArgumentError("Can't have a negative input length!");
+    }
+
+    int blockSize = _cipher.blockSize;
+    int gapLen = blockSize - _bufOff;
+
+    if (len > gapLen) {
+      _buf.setRange(_bufOff, _bufOff + gapLen, inp.sublist(inOff));
+
+      _cipher.processBlock(_buf, 0, _mac, 0);
+
+      _bufOff = 0;
+      len -= gapLen;
+      inOff += gapLen;
+
+      while (len > blockSize) {
+        _cipher.processBlock(inp, inOff, _mac, 0);
+
+        len -= blockSize;
+        inOff += blockSize;
+      }
+    }
+
+    _buf.setRange(_bufOff, _bufOff + len, inp.sublist(inOff));
+
+    _bufOff += len;
+  }
+
+  @override
+  int doFinal(Uint8List out, int outOff) {
+    int blockSize = _cipher.blockSize;
+
+    if (_padding == null) {
+      //
+      // pad with zeroes
+      //
+      while (_bufOff < blockSize) {
+        _buf[_bufOff] = 0;
+        _bufOff++;
+      }
+    } else {
+      if (_bufOff == blockSize) {
+        _cipher.processBlock(_buf, 0, _mac, 0);
+        _bufOff = 0;
+      }
+
+      _padding.addPadding(_buf, _bufOff);
+    }
+
+    _cipher.processBlock(_buf, 0, _mac, 0);
+
+    out.setRange(outOff, outOff + _macSize, _mac);
+
+    reset();
+
+    return _macSize;
+  }
+
+  /// Reset the mac generator.
+  @override
+  void reset() {
+    // clean the buffer.
+    for (int i = 0; i < _buf.length; i++) {
+      _buf[i] = 0;
+    }
+
+    _bufOff = 0;
+
+    // reset the underlying cipher.
+    _cipher.reset();
+
+    _cipher.init(true, _params);
```

Comment:
Again, this seems weird to me, but storing `_params` and re-initializing the key seems to be necessary. Maybe there's a better place to put it (e.g. just after every `reset` call site?)